os/io.c

```diff
@@ -680,17 +701,22 @@ static bool OutputBufferMakeRoom(ClientPtr who, OsCommPtr oc, size_t sz)
     const size_t padsize = padding_for_int32(sz);
     const size_t needed = sz + padsize;
 
-    ConnectionOutputPtr oco = oc->output;
+    if (oc->output) {
+        /* check whether it already fits into buffer */
+        if (oc->output->count + needed <= oc->output->size)
+            return true;
 
-    /* check whether it already fits into buffer */
-    if (oco->count + needed <= oco->size)
-        return true;
+        /* try flushing the buffer */
+        int ret = FlushClient(who, oc);
+        if (ret == -1) /* client was aborted */
+            return false;
+    }
 
-    /* try flushing the buffer */
-    int ret = FlushClient(who, oc);
-    if (ret == -1) /* client was aborted */
+    if (!OutputEnsureBuffer(who, oc))
```

Comment:
I'm not 100% follow the logic here, but wouldn't it make sense to have the `OutputEnsureBuffer` call before the `if (oc->output) ...` (line 704 currently)?

Reply:
Hmm, if `oc->output` is non-null then `OutputEnsureBuffer()` would do nothing. So the code inside that `if` would execute as usual (we don't even have to check for null in this case). 

Do we really need to call `OutputEnsureBuffer()` here (maybe do it on the calling side only)? This method looks like it was intended only for flush/realloc if needed.

Reply:
We've just called FlushClient() here, and that may kill the buffer.
If it wouldn't then we explicitly need to kill it on client shutdown (which currently is relying on FlushClient() doing that).